### PR TITLE
Allow truncating fraction part in cast(string as integer)

### PR DESCRIPTION
**Pull request by marin-ma**
Summary: SparkSQL returns 1 on cast("1.8" as int) and returns -1 on cast("-1.8" as int), while Velox throws. This fix make Velox return the same as SparkSQL. 

This change only affects setting cast_to_int_by_truncate=true. So it doesn't affect Presto CAST. Currently Velox will throw for Presto cast(string as integer), which matches presto's behavior. 

Fixes: https://github.com/facebookincubator/velox/issues/6110

### CODE REVIEW

**Comment by marin-ma** [comment]
cc: @PHILO-HE

**Comment by marin-ma** [comment]
@bikramSingh91 Could you help to review? Thanks!

**Comment by marin-ma** [comment]
@pedroerp Could you please help review?

**Comment by pedroerp** [comment]
@marin-ma looks like this change would change the behavior in the entire library (not just for Spark). Can you double check what's the behavior expected in Presto?

**Comment by marin-ma** [comment]
> @marin-ma looks like this change would change the behavior in the entire library (not just for Spark). Can you double check what's the behavior expected in Presto?

@kagamiori Could you help double check? Here's what I got from presto:

image

**Comment by jinchengchenghh** [comment]
The behavior is different.

Spark result:
```
Spark master: local[4], Application Id: local-1692734049820
spark-sql> select cast('1.8' as int);
1
Time taken: 2.955 seconds, Fetched 1 row(s)
spark-sql> select cast('-1.8' as int);
-1
Time taken: 0.113 seconds, Fetched 1 row(s)
```

Presto result：
```
presto:default> select cast('1.8' as int);
Query 20230822_195358_00006_sqyi8 failed: Cannot cast '1.8' to INT

presto:default> select cast('-1.8' as int);
Query 20230822_195401_00007_sqyi8 failed: Cannot cast '-1.8' to INT
```

**Comment by marin-ma** [comment]
> The behavior is different.
> 
> Spark result:
> 
> ```
> Spark master: local[4], Application Id: local-1692734049820
> spark-sql> select cast('1.8' as int);
> 1
> Time taken: 2.955 seconds, Fetched 1 row(s)
> spark-sql> select cast('-1.8' as int);
> -1
> Time taken: 0.113 seconds, Fetched 1 row(s)
> ```
> 
> Presto result：
> 
> ```
> presto:default> select cast('1.8' as int);
> Query 20230822_195358_00006_sqyi8 failed: Cannot cast '1.8' to INT
> 
> presto:default> select cast('-1.8' as int);
> Query 20230822_195401_00007_sqyi8 failed: Cannot cast '-1.8' to INT
> ```

Yes. As described in the issue. Spark sql truncates the string while presto rounds the string.

**Comment by bikramSingh91** [comment]
@marin-ma 
As @jinchengchenghh mentioned, I also confirmed that presto (JAVA and native both) latest does not support casting strings that contain real/double to integers. Instead of rounding, it is throwing an error. Is there a query config that we need to turn on to support this?

If Presto indeed does not support this conversion, I am not sure if we should to piggyback onto the existing 'cast_to_int_by_truncate' query config. @mbasmanova do you have any context on why this config was added? was it added to support spark cases similar to how a new config 'cast_string_to_date_non_standard' is being added here #5844 to support spark date conversions.
Also, I am wondering if there is a another way to add this kind of additional functionality which is specific to spark instead of using query configs that dont communicate this context directly.

File: velox/type/Conversions.h [hunk 1]
```diff
@@ -102,6 +102,10 @@ struct Converter<
       }
       if (negative) {
         for (; index < len; index++) {
+          // Ignore the fractional part for floating point/decimal.
```
Comment (bikramSingh91) [inline velox/type/Conversions.h:105]: this check would allow anything after a decimal to be ignored, which means "1.abc" will be valid. Is this expected?

**Comment by mbasmanova** [comment]
> cast_to_int_by_truncate

@bikramSingh91 This config is being used by Spark. @rui-mo may have more context.

> cast_string_to_date_non_standard

I wonder about this new config name. "non_standard" is not very informative. Can we come up with a more descriptive name? 

I'm seeing that Spark's CAST behavior is different from Presto in multiple edge cases. It would be nice to first collect all the differences together into a single GitHub issue / document so we can figure out a good way to support all of them rather than solve each case in isolation.

**Comment by rui-mo** [comment]
> cast_to_int_by_truncate

@mbasmanova Yes, Gluten enables this config to get matched result with Spark, e.g., when casting from double integer. 

> It would be nice to first collect all the differences together into a single GitHub issue / document so we can figure out a good way to support all of them rather than solve each case in isolation.

@PHILO-HE I remember you collected these cases, and do we have an issue on that?

**Comment by marin-ma** [comment]
> @marin-ma As @jinchengchenghh mentioned, I also confirmed that presto (JAVA and native both) latest does not support casting strings that contain real/double to integers. Instead of rounding, it is throwing an error. Is there a query config that we need to turn on to support this?
> 
> If Presto indeed does not support this conversion, I am not sure if we should to piggyback onto the existing 'cast_to_int_by_truncate' query config. @mbasmanova do you have any context on why this config was added? was it added to support spark cases similar to how a new config 'cast_string_to_date_non_standard' is being added here #5844 to support spark date conversions. Also, I am wondering if there is a another way to add this kind of additional functionality which is specific to spark instead of using query configs that dont communicate this context directly.

@bikramSingh91 Thanks for the double check. Let me remove the presto part from the issue description.

**Comment by PHILO-HE** [comment]
> > It would be nice to first collect all the differences together into a single GitHub issue / document so we can figure out a good way to support all of them rather than solve each case in isolation.
> 
> @PHILO-HE I remember you collected these cases, and do we have an issue on that?

We recognized a few cases and posted them here: https://github.com/facebookincubator/velox/issues/4876#issuecomment-1554537310

File: velox/type/Conversions.h (hunk 1 above)
Comment (marin-ma) [reply velox/type/Conversions.h]: Thanks! "1.abc" is invalid. Fixed in the latest commit and added test cases.

**Comment by marin-ma** [comment]
@mbasmanova @bikramSingh91 Could you help review again? Thanks!

**Review by mbasmanova — COMMENTED** [review]
@marin-ma Does this change affect Presto CAST as well? If so, does the new behavior match Presto?

Please, update PR title to capture the change in this PR. Currently, it sounds like a problem statement: Cast behaves differently. Let's change it to describe the solution.

**Review by mbasmanova — COMMENTED** [review]
Please, update CAST documentation as well. https://facebookincubator.github.io/velox/functions/presto/conversion.html

**Comment by marin-ma** [comment]
@mbasmanova Could you help review again? Thanks!

**Review by mbasmanova — COMMENTED** [review]
@bikramSingh91 Bikram, would you help review and merge this PR?

**Comment by jinchengchenghh** [comment]
> > @marin-ma Does this change affect Presto CAST as well? If so, does the new behavior match Presto?
> > Please, update PR title to capture the change in this PR. Currently, it sounds like a problem statement: Cast behaves differently. Let's change it to describe the solution.
> 
> This change only affects setting `cast_to_int_by_truncate=true`. So it doesn't affect Presto CAST. Currently Velox will throw for Presto cast(string as integer), which matches presto's behavior. The previous result that I got from presto was wrong. @jinchengchenghh and @bikramSingh91 helped to correct and confirm that.

Please update this in the PR description.

File: velox/expression/tests/CastExprTest.cpp [hunk 2]
```diff
@@ -798,6 +797,15 @@ TEST_F(CastExprTest, primitiveValidCornerCases) {
   setCastIntByTruncate(true);
   // To integer.
   {
+    // Valid strings.
+    testCast<std::string, int8_t>("tinyint", {"1.2"}, {1}, false);
+    testCast<std::string, int8_t>("tinyint", {"-1.8"}, {-1}, false);
+    testCast<std::string, int8_t>("tinyint", {"1."}, {1}, false);
```
Comment (bikramSingh91) [inline velox/expression/tests/CastExprTest.cpp:803]: cases from L803-L807 are very unexpected and not obvious. Can you mention that these are valid cases as well in. the description for `cast_to_int_by_truncate`

File: velox/expression/tests/CastExprTest.cpp [hunk 3]
```diff
@@ -798,6 +797,15 @@ TEST_F(CastExprTest, primitiveValidCornerCases) {
   setCastIntByTruncate(true);
   // To integer.
   {
+    // Valid strings.
+    testCast<std::string, int8_t>("tinyint", {"1.2"}, {1}, false);
```
Comment (bikramSingh91) [inline velox/expression/tests/CastExprTest.cpp:801]: can you also add a cases with a larger fractional part? like 1.23444 or ".2355", etc

**Comment by marin-ma** [comment]
@bikramSingh91 Could you help to review again? Thanks!

**Review by bikramSingh91 — APPROVED** [review]

**Comment by marin-ma** [comment]
@bikramSingh91 Rebased to resolve the conflicts. Can this one get merged?

**Comment by bikramSingh91** [comment]
> @bikramSingh91 Rebased to resolve the conflicts. Can this one get merged?

Trying to get this merged today, just need to wait for internal pre-commit tests to pass again after pulling latest changes to the PR today.

**Pull request closed without merging**